rollup/cmd/gas_oracle/app/app.go

```diff
@@ -106,9 +97,31 @@ func action(ctx *cli.Context) error {
 		}
 	})
 
-	// Start l1relayer process
-	go utils.Loop(subCtx, 10*time.Second, l1relayer.ProcessGasPriceOracle)
-	go utils.Loop(subCtx, 2*time.Second, l2relayer.ProcessGasPriceOracle)
+	// Goroutines that may send transactions periodically.
+	go utils.Loop(loopCtx, 10*time.Second, l1relayer.ProcessGasPriceOracle)
+	go utils.Loop(loopCtx, 2*time.Second, l2relayer.ProcessGasPriceOracle)
+
+	defer func() {
+		log.Info("Graceful shutdown initiated")
+
+		// Prevent new transactions by canceling the loop context.
+		loopCancel()
+
+		// Close relayers to ensure all pending transactions are processed.
+		// This includes any in-flight transactions that have not yet been confirmed.
+		l1relayer.Close()
+		l2relayer.Close()
+
+		// Halt confirmation signal handling by canceling the instance context.
+		instanceCancel()
+
+		// Close the database connection.
+		if err = database.CloseDB(db); err != nil {
+			log.Error("Failed to close database connection", "error", err)
+		}
+
+		log.Info("Graceful shutdown done")
```

Comment:
Important to node that `docker stop` and `docker restart` will kill the process if it doesn't terminate within a certain period. We'll also need to adjust this by passing a flag with a longer period.

Reply:
Yes, like this: https://github.com/scroll-tech/testnet/pull/840 (currently only configured for `l1geth`).